src/ripple/conditions/impl/Fulfillment.cpp

```diff
@@ -0,0 +1,216 @@
+    // what the condition allows.
+    if (f.payloadSize() > c.maxFulfillmentLength)
+        return false;
+
+    return f.condition() == c;
+}
+
+bool
+validate (
+    Fulfillment const& f,
+    Condition const& c,
+    Slice m)
+{
+    return fulfills (f, c) && f.validate (m);
+}
+
+bool
+validateTrigger (
+    Fulfillment const& f,
+    Condition const& c)
+{
+    return validate (f, c, {});
+}
+
+std::unique_ptr<Fulfillment>
+loadFulfillment (std::uint16_t type, Slice payload)
+{
+    std::unique_ptr<Fulfillment> p;
+
+    switch (type)
+    {
+    case condition_hashlock:
+        p = std::make_unique<PreimageSha256>();
+        break;
+
+    case condition_prefix_sha256:
+        p = std::make_unique<PrefixSha256>();
+        break;
+
+    case condition_rsa_sha256:
+        p = std::make_unique<RsaSha256>();
+        break;
+
+    case condition_ed25519:
+        p = std::make_unique<Ed25519>();
+        break;
+
+    default:
+        throw std::domain_error (
+            "Unknown cryptocondition type " +
+                std::to_string (type));
+    }
+
+    // If the payload can't be parsed, the load should
+    // fail.
+    if (p && !p->parsePayload(payload))
+        p.reset();
+
+    return p;
+}
+
+// Parse a condition from its string form
+std::unique_ptr<Fulfillment>
+loadFulfillment (std::string const& s)
+{
+    // CHECKME: custom parser maybe? probably faster but
+    // more work and probability of error.
+
+    // TODO: use two regex: one that accepts anything the
+    // standard supports and one which accepts only what
+    // we support. Parse with both for improved errors?
+    static boost::regex const re_current (
+            "^"                            // start of line
+            "cf:"                          // 'cf' for fulfillment
+            "([1-9a-f][0-9a-f]{0,3}|0):"   // type
+            "([a-zA-Z0-9_-]*)"             // fulfillment payload (base64url)
+            "$"                            // end of line
+        , boost::regex_constants::optimize
+    );
+
+    try
+    {
+        boost::smatch match;
+
+        if (!boost::regex_match (s, match, re_current))
+            return nullptr;
+
+        std::uint16_t const type =
+            parse_hexadecimal<std::uint16_t>(match[1]);
+
+        auto payload = base64url_decode (match[2]);
+
+        if (payload.size() > maxSupportedFulfillmentLength)
+            return nullptr;
+
+        return loadFulfillment (type, makeSlice (payload));
+    }
+    catch (std::exception const& x)
+    {
+        return nullptr;
+    }
+}
+
+std::unique_ptr<Fulfillment>
+loadFulfillment (Slice s)
+{
+    try
+    {
+        auto iter = s.cbegin();
+
+        std::uint16_t type;
+        std::size_t len;
+
+        std::tie (iter, type) =
+            oer::decode_integer<std::uint16_t> (
+                iter, s.cend());
+
+        std::tie (iter, len) =
+            oer::decode_length(
+                iter, s.cend());
```

Comment:
`decode_length` dereferences `*iter` immediately. We should check for a short buffer before this call.